app/grandchallenge/algorithms/views.py

```diff
@@ -339,22 +341,47 @@ def get_context_data(self, *args, **kwargs):
 class AlgorithmImageMarkDesired(
     LoginRequiredMixin, ObjectPermissionRequiredMixin, View
 ):
-    permission_required = "algorithms.change_algorithmimage"
+    permission_required = "algorithms.change_algorithm"
     raise_exception = True
 
     @cached_property
     def algorithm(self) -> Algorithm:
-        return get_object_or_404(Algorithm, slug=self.kwargs["slug"])
+        return self.algorithm_image.algorithm
 
     @cached_property
-    def algorithm_image(self) -> Algorithm:
+    def algorithm_image(self) -> AlgorithmImage:
         return get_object_or_404(AlgorithmImage, pk=self.kwargs["pk"])
 
     def get_permission_object(self):
-        return self.algorithm_image
+        return self.algorithm
+
+    def post(self, request, *args, **kwargs):
+        if not self.algorithm_image.is_manifest_valid:
+            messages.error(
+                request, "Cannot mark an invalid image as active image."
+            )
+            return HttpResponseRedirect(
+                self.algorithm_image.get_absolute_url()
+            )
+        if not self.algorithm_image.can_execute:
+            try:
+                upload_to_registry_and_sagemaker(self.algorithm_image)
```

Comment:
This needs to be done asynchronously. Uploading a file from S3 to ECR can take up to 20 minutes as these images are often massive. You need to be really careful here that this is only being done once per algorithm image or algorithm.

Reply:
I'm unsure how to make sure this only runs once per image. I would need to be able to check whether there is already an sync task scheduled or running for this image, right?

Reply:
Yes, much like we do for the initial importing.